sdk/include/opentelemetry/sdk/trace/processor.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/span.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/trace/span.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+/**
+ * Span processor allow hooks for span start and end method invocations.
+ *
+ * Built-in span processors are responsible for batching and conversion of
+ * spans to exportable representation and passing batches to exporters.
+ */
+class SpanProcessor
+{
+public:
+  virtual ~SpanProcessor() = default;
+
+  // OnStart is called when a span is started.
+  virtual void OnStart(opentelemetry::trace::Span &span) noexcept = 0;
+
+  // OnEnd is called when a span is ended.
+  virtual void OnEnd(
+      nostd::shared_ptr<opentelemetry::sdk::trace::Recordable> &recordable) noexcept = 0;
```

Comment:
Why a shared_ptr instead of `std::unique_ptr<opentelemetry::sdk::trace::Recordable>&&`?

Reply:
Because a tracer can have multiple processors, and a processor might store recordables in a batch. I think we don't want to make unique copies of recordables before passing them on to processors.

Reply:
How does this work then if you're using Recordable types?

I'd prefer a design where each Exporter-SpanProcessor has a single unique Recordable. I don't think it's necessary to do any copying of the Recordables.

Reply:
I see, a span carries a recordabe for each processor.

I changed this to use `unique_ptr`.